File: pkg/snowblock/task/link/link.go
// Copyright (C) 2017-present Arctic Ice Studio <development@arcticicestudio.com>
// Copyright (C) 2017-present Sven Greb <development@svengreb.de>
//
// Project:    snowsaw
// Repository: https://github.com/arcticicestudio/snowsaw
// License:    MIT

// Author: Arctic Ice Studio <development@arcticicestudio.com>
// Author: Sven Greb <development@svengreb.de>
// Since: 0.4.0

// Package link provides a task runner implementation to create symbolic links for files and directories.
package link

import (
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/fatih/color"
	"github.com/mitchellh/mapstructure"

	"github.com/arcticicestudio/snowsaw/pkg/api/snowblock"
	"github.com/arcticicestudio/snowsaw/pkg/prt"
	"github.com/arcticicestudio/snowsaw/pkg/util/filesystem"
)

const (
	// DefaultHostName is the name for host mappings that will apply to all host.
	// To prevent possible collisions with actual host names, it is a single minus character.
	// As defined in the specification this is not a valid hostname since the name should not start or end with a minus.
	// See "RFC 1123" and https://en.wikipedia.org/wiki/Hostname#Restrictions_on_valid_hostnames for more details about
	// restrictions and valid names.
	DefaultHostName = "-"
)

// Link is a task runner to create symbolic links for files and directories.
type Link struct {
	config           *config
	destAbsPath      string
	destPath         string
	snowblockAbsPath string
	srcAbsPath       string
}

type config struct {
	Create   bool              `json:"create" yaml:"create"`
	Force    bool              `json:"force" yaml:"force"`
	Hosts    map[string]string `json:"hosts,flow" yaml:"hosts,flow"`
	Path     string            `json:"path" yaml:"path"`
	Relative bool              `json:"relative" yaml:"relative"`
	Relink   bool              `json:"relink" yaml:"relink"`
}

// GetTaskName returns the name of the task this runner can process.
func (l Link) GetTaskName() string {
	return "link"
}

// Run processes a task using the given task instructions.
// The snowblockAbsPath parameter is the absolute path of the snowblock used as contextual information.
func (l *Link) Run(configuration snowblock.TaskConfiguration, snowblockAbsPath string) error {
	l.snowblockAbsPath = snowblockAbsPath

	// Try to convert given task configurations...
	configMap, ok := configuration.(map[string]interface{})
	if !ok {
		prt.Debugf("invalid link configuration type: %s", color.RedString("%T", configuration))
		return errors.New("malformed link configuration")
	}

	// ...and handle the possible types.
	for destPath, configData := range configMap {
		l.destAbsPath = ""
		l.srcAbsPath = ""

		switch configType := configData.(type) {
		// Handle JSON `null` value configurations used to omit duplicate definitions when the source path equals the
		// destination path.
		// Uses the base name of the destination path and trims a leading dot character if present.
		case nil:
			sourceBaseName := strings.TrimPrefix(filepath.Base(destPath), ".")
			l.config = &config{Path: sourceBaseName}
			l.destPath = destPath
			if execErr := l.execute(); execErr != nil {
				return execErr
			}

		// Handle JSON `object` configurations used to define more link options.
		// Uses the base name of the destination path with leading dot character trimmed if path is not specified.
		case map[string]interface{}:
			c := new(config)
			if err := mapstructure.Decode(configType, &c); err != nil {
				return err
			}
			l.destPath = destPath
			if c.Path == "" {
				c.Path = strings.TrimPrefix(filepath.Base(destPath), ".")
			}
			l.config = c
			if execErr := l.execute(); execErr != nil {
				return execErr
			}

		// Handle JSON `string` configurations used to only specify the source path.
		case string:
			l.config = &config{Path: configType}
			l.destPath = destPath
			if execErr := l.execute(); execErr != nil {
				return execErr
			}

		// Reject invalid or unsupported JSON data structures.
		default:
			prt.Debugf("unsupported destination type: %s", color.RedString("%T", configType))
			return fmt.Errorf("unsupported link configuration: %s", color.CyanString(destPath))
		}
	}

	return nil
}

func (l *Link) execute() error {
	// Check if the current and/or default host is listed in the target mapping, otherwise stop processing.
	isTargetHost, hostCheckErr := l.isTargetHost()
	if hostCheckErr != nil {
		return hostCheckErr
	}
	if !isTargetHost {
		return nil
	}

	// Dissolve the source to an absolute path.
	srcAbsPath, srcToAbsPathErr := filepath.Abs(filepath.Join(l.snowblockAbsPath, l.config.Path))
	if srcToAbsPathErr != nil {
		return srcToAbsPathErr
	}
	l.srcAbsPath = srcAbsPath

	// Fail fast if the source node does not exist.
	if sourceNodeExistsErr := l.checkSourceNode(); sourceNodeExistsErr != nil {
		return sourceNodeExistsErr
	}

	// Expand the destination path to dissolve environment variables and special characters like tilde...
	expDestPath, pathExpandErr := filesystem.ExpandPath(l.destPath)
	if pathExpandErr != nil {
		return pathExpandErr
	}

	if !filepath.IsAbs(expDestPath) {
		l.destAbsPath = filepath.Join(l.snowblockAbsPath, expDestPath)
	} else {
		l.destAbsPath = expDestPath
	}

	destNodeExists, nodeExistErr := filesystem.NodeExists(l.destAbsPath)
	if nodeExistErr != nil {
		return nodeExistErr
	}
	// Check if the destination node already exists,...
	if destNodeExists {
		isSymlink, symlinkCheckErr := filesystem.IsSymlink(l.destAbsPath)
		if symlinkCheckErr != nil {
			return symlinkCheckErr
		}
		// ...evaluate if it is a symbolic link,...
		if isSymlink {
			symlinkDest, symlinkReadErr := os.Readlink(l.destAbsPath)
			if symlinkReadErr != nil {
				return symlinkReadErr
			}
			symlinkDestAbs, symlinkDestAbsErr := filepath.Abs(symlinkDest)
			if symlinkDestAbsErr != nil {
				return symlinkDestAbsErr
			}

			// ...and continue with processing when running in relinking mode,...
			if l.config.Relink {
				prt.Warnf("%s already existing symbolic link: %s",
					color.YellowString("Relinking"), color.CyanString(l.destAbsPath))
				if removeErr := os.Remove(l.destAbsPath); removeErr != nil {
					return removeErr
				}
				if parentDirErr := l.handleParentDirStructure(); parentDirErr != nil {
					return parentDirErr
				}
				if symlinkCreationError := l.createSymbolicLink(); symlinkCreationError != nil {
					return symlinkCreationError
				}
				return nil
			}

			// ...or stop processing when it already links to the correct destination,...
			if symlinkDestAbs == l.srcAbsPath {
				prt.Infof("Skipped already existing link: %s", color.CyanString(l.destAbsPath))
				return nil
			}

			// ...otherwise only if force linking is enabled.
			if l.config.Force {
				prt.Warnf("%s of already existing symbolic link: %s",
					color.YellowString("Forced linking"), color.CyanString(l.destAbsPath))
				if removeErr := os.Remove(l.destAbsPath); removeErr != nil {
					return removeErr
				}
				if parentDirErr := l.handleParentDirStructure(); parentDirErr != nil {
					return parentDirErr
				}
				if symlinkCreationError := l.createSymbolicLink(); symlinkCreationError != nil {
					return symlinkCreationError
				}
				return nil
			}

			return fmt.Errorf("symbolic link already exists: %s ← %s", symlinkDest, l.destAbsPath)
		}

		// Always process the task in force mode when the destination is an already existing file or directory,...
		if l.config.Force {
			prt.Warnf("%s of already existing symbolic link: %s",
				color.YellowString("Forced linking"), color.CyanString(l.destAbsPath))
			if removeErr := os.Remove(l.destAbsPath); removeErr != nil {
				return removeErr
			}
			if parentDirErr := l.handleParentDirStructure(); parentDirErr != nil {
				return parentDirErr
			}
			if symlinkCreationError := l.createSymbolicLink(); symlinkCreationError != nil {
				return symlinkCreationError
			}
			return nil
		}

		return fmt.Errorf("file or directory already exists: %s", l.destAbsPath)
	}

	// ...otherwise only when all previous conditions are not met.
	if parentDirErr := l.handleParentDirStructure(); parentDirErr != nil {
		return parentDirErr
	}
	if symlinkCreateErr := l.createSymbolicLink(); symlinkCreateErr != nil {
		return symlinkCreateErr
	}

	return nil
}

// checkSourceNode checks if the source node at the given path exists, otherwise returns the corresponding error.
func (l *Link) checkSourceNode() error {
	sourceNodeExists, err := filesystem.NodeExists(l.srcAbsPath)
	if err != nil {
		return err
	}
	if !sourceNodeExists {
		return fmt.Errorf("no such file or directory: %s", l.config.Path)
	}

	return nil
}

// createSymbolicLink creates the symbolic link based on the value of the task option that allows to use relative
// instead of absolute paths.
// If any error occurs it will be returned, otherwise returns nil.
func (l *Link) createSymbolicLink() error {
	if l.config.Relative {
		srcRelPath, srcRelPathErr := filepath.Rel(filepath.Dir(l.destAbsPath), l.srcAbsPath)
		if srcRelPathErr != nil {
			return fmt.Errorf("could not dissolve path of source relative to destination directory: %v", srcRelPathErr)
		}

		if relSymlinkErr := os.Symlink(srcRelPath, l.destAbsPath); relSymlinkErr != nil {
			return relSymlinkErr
		}
		prt.Infof("Created relative symbolic link: %s → %s", color.CyanString(l.srcAbsPath), color.BlueString(l.srcAbsPath))
		return nil
	}

	if symlinkErr := os.Symlink(l.srcAbsPath, l.destAbsPath); symlinkErr != nil {
		return symlinkErr
	}
	prt.Infof("Created symbolic link: %s → %s", color.BlueString(l.destAbsPath), color.CyanString(l.srcAbsPath))
	return nil
}

// handleParentDirStructure checks if the required parent directory structure for the symbolic links exists,
// otherwise creates it if the corresponding task option has been specified.
// If any error occurs it will be returned, otherwise returns nil.
func (l *Link) handleParentDirStructure() error {
	destParentDirs := filepath.Dir(l.destAbsPath)
	destParentDirsExist, nodeExistErr := filesystem.DirExists(destParentDirs)
	if nodeExistErr != nil {
		return nodeExistErr
	}
	if !destParentDirsExist {
		if l.config.Create {
			if mkdirErr := os.MkdirAll(destParentDirs, os.ModePerm); mkdirErr != nil {
				return mkdirErr
			}
			prt.Debugf("Created parent directory structure: %s", destParentDirs)
		} else {
			return fmt.Errorf("no such directory: %s", destParentDirs)
		}
	}

	return nil
}

// isTargetHost checks if the current and/or default host is listed in the target mapping.
// It returns the host specific source path, otherwise if an error occurs an empty string along with the error.
func (l *Link) isTargetHost() (bool, error) {
	if len(l.config.Hosts) > 0 {
		hostname, err := os.Hostname()
		if err != nil {
			return false, fmt.Errorf("failed to determine hostname: %v", err)
		}
		sourcePath, isTargetHost := l.config.Hosts[hostname]
		sourcePathDefaultHost, isDefaultTargetHost := l.config.Hosts[DefaultHostName]
		if !isTargetHost && !isDefaultTargetHost {
			prt.Debugf("Skipped host specific link not matching current host %s: %s",
				color.BlueString(hostname), color.CyanString(l.destPath))
			return false, nil
		}

		// Use the default target host if specified...
		if isDefaultTargetHost {
			prt.Debugf("Found host mapping for default target: %s", color.CyanString(sourcePathDefaultHost))
			l.config.Path = sourcePathDefaultHost
		}
		// ...and override when exact host name has also been specified.
		if isTargetHost {
			prt.Debugf("Using source path for exact host name match %s: %s",
				color.BlueString(hostname), color.CyanString(sourcePath))
			l.config.Path = sourcePath
		}

		return true, nil
	}

	if l.config.Path != "" {
		return true, nil
	}

	return false, nil
}

File: snowsaw/plugins/link.py
# -*- coding: utf-8 -*-

import os
import shutil
import snowsaw
from socket import gethostname


class Link(snowsaw.Plugin):
    """
    Core plugin to symbolically link dotfiles.
    """
    _directive = "link"

    def can_handle(self, directive):
        return directive == self._directive

    def handle(self, directive, data):
        if directive != self._directive:
            raise ValueError("Core plugin \"Link\" cannot handle the directive \"{}\"".format(directive))
        return self._process_links(data)

    def _process_links(self, links):
        """
        Processes specified links.

        :param links: The links to process
        :return: True if the links have been processed successfully, False otherwise
        """
        success = True
        defaults = self._context.defaults().get("link", {})
        hostname = gethostname()
        for destination, source in links.items():
            destination = os.path.expandvars(destination)
            relative = defaults.get("relative", False)
            force = defaults.get("force", False)
            relink = defaults.get("relink", False)
            create = defaults.get("create", False)
            hosts = defaults.get("hosts", {})
            if isinstance(source, dict):
                relative = source.get("relative", relative)
                force = source.get("force", force)
                relink = source.get("relink", relink)
                create = source.get("create", create)
                path = self._default_source(destination, source.get("path"))
                hosts = source.get("hosts", hosts)
            else:
                path = self._default_source(destination, source)
            path = os.path.expandvars(os.path.expanduser(path))

            if hosts:
                if hostname in hosts:
                    path = os.path.expandvars(os.path.expanduser(hosts.get(hostname)))
                elif "-" in hosts:
                    path = os.path.expandvars(os.path.expanduser(hosts.get("-")))
                    self._log.lowinfo("Applying default link {} -> {}".format(destination, os.path.join(self._context.snowblock_dir(),path)))
                else:
                    for host in hosts.items():
                        self._log.lowinfo("Skipped host specific link {} -> {}".format(destination, os.path.join(self._context.snowblock_dir(), host[1])))
                    continue

            if not self._exists(os.path.join(self._context.snowblock_dir(), path)):
                success = False
                self._log.warning("Nonexistent target {} -> {}".format(destination, path))
                continue
            if create:
                success &= self._create(destination)
            if force or relink:
                success &= self._delete(path, destination, relative, force)
            success &= self._link(path, destination, relative)
        if success:
            self._log.info("=> All links have been set up")
        else:
            self._log.error("Some links were not successfully set up")
        return success

    def _default_source(self, destination, source):
        """
        Sets the default source if the value is empty, the configured source otherwise.

        If the source for a link is null, it uses the basename of the destination.
        Leading dot characters in the basename will be stripped.

        Allows simplified configuration files by avoiding unnecessary duplicate values.

        :param destination: The link destination
        :param source: The link source
        :return: The source string
        """
        if source is None:
            basename = os.path.basename(destination)
            if basename.startswith('.'):
                return basename[1:]
            else:
                return basename
        else:
            return source

    def _is_link(self, path):
        """
        Checks if the specified path is a symbolic link.

        :param path: The path to check
        :return: True if the path is a symbolic link
        """
        return os.path.islink(os.path.expanduser(path))

    def _link_destination(self, path):
        """
        Gets the destination of the specified symbolic link.

        :param path: The symbolic link to get the destination of
        :return: The symbolic link destination
        """
        path = os.path.expanduser(path)
        return os.readlink(path)

    def _exists(self, path):
        """
        Checks if the specified path exists.

        :param path: The path to check
        :return: True if the path exists, False for broken symbolic links otherwise
        """
        path = os.path.expanduser(path)
        return os.path.exists(path)

    def _create(self, path):
        """
        Creates a symbolic link to the specified path.

        :param path: The path to the symbolic link to create
        :return: True if the symbolic link has been created successfully, False otherwise
        """
        success = True
        parent = os.path.abspath(os.path.join(os.path.expanduser(path), os.pardir))
        if not self._exists(parent):
            try:
                os.makedirs(parent)
            except OSError:
                self._log.warning("Failed to create directory \"{}\"".format(parent))
                success = False
            else:
                self._log.lowinfo("Creating directory \"{}\"".format(parent))
        return success

    def _delete(self, source, path, relative, force):
        """
        Deletes the specified path.

        :param source: The link source
        :param path: The path to delete
        :param relative: Flag to indicate if the specified parameters are relative
        :param force: Flag to indicate if the deletion should be forced
        :return: True if the path has been deleted successfully, False otherwise
        """
        success = True
        source = os.path.join(self._context.snowblock_dir(), source)
        fullpath = os.path.expanduser(path)
        if relative:
            source = self._relative_path(source, fullpath)
        if (self._is_link(path) and self._link_destination(path) != source) or (self._exists(path) and not self._is_link(path)):
            removed = False
            try:
                if os.path.islink(fullpath):
                    os.unlink(fullpath)
                    removed = True
                elif force:
                    if os.path.isdir(fullpath):
                        shutil.rmtree(fullpath)
                        removed = True
                    else:
                        os.remove(fullpath)
                        removed = True
            except OSError:
                self._log.warning("Failed to remove {}".format(path))
                success = False
            else:
                if removed:
                    self._log.lowinfo("Removing {}".format(path))
        return success

    def _relative_path(self, source, destination):
        """
        Gets the relative path to get to the source path from the destination path.

        :param source: The source path
        :param destination: The destination path
        :return: The relative path
        """
        destination_dir = os.path.dirname(destination)
        return os.path.relpath(source, destination_dir)

    def _link(self, source, link_name, relative):
        """
        Links the specified link_name to the source.

        :param source: The source path to get linked
        :param link_name: The name of the link to link
        :return: True if the link has been linked successfully, False otherwise
        """
        success = False
        destination = os.path.expanduser(link_name)
        absolute_source = os.path.join(self._context.snowblock_dir(), source)
        if relative:
            source = self._relative_path(absolute_source, destination)
        else:
            source = absolute_source
        if not self._exists(link_name) and self._is_link(link_name) and self._link_destination(link_name) != source:
            self._log.warning("Invalid link {} -> {}".format(link_name, self._link_destination(link_name)))
        elif not self._exists(link_name) and self._exists(absolute_source):
            try:
                os.symlink(source, destination)
            except OSError:
                self._log.warning("Linking failed {} -> {}".format(link_name, source))
            else:
                self._log.lowinfo("Creating link {} -> {}".format(link_name, source))
                success = True
        elif self._exists(link_name) and not self._is_link(link_name):
            self._log.warning("{} already exists but is a regular file or directory".format(link_name))
        elif self._is_link(link_name) and self._link_destination(link_name) != source:
            self._log.warning("Incorrect link {} -> {}".format(link_name, self._link_destination(link_name)))
        elif not self._exists(absolute_source):
            if self._is_link(link_name):
                self._log.warning("Nonexistent target {} -> {}".format(link_name, source))
            else:
                self._log.warning("Nonexistent target for {} : {}".format(link_name, source))
        else:
            self._log.lowinfo("Link already exists {} -> {}".format(link_name, source))
            success = True
        return success
